### PR TITLE
[FLINK-13253][jdbc] Deadlock may occur in JDBCUpsertOutputFormat

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

In close, it await the flush scheduler terminal, but it hold the lock of JDBCUpsertOutputFormat instance, maybe the async thread is waiting for this lock in the flush method, so there might be a deadlock here.

First, it should not await scheduler terminal, because it has flushed all data to jdbc, what we should do is let async thread quit.
Second, we should add lock outside the closed check in the flusher, in this way, we can ensure async thread secure exiting.

## Verifying this change

Manual test

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java [hunk 1]
```diff
@@ -116,13 +116,15 @@ public void open(int taskNumber, int numTasks) throws IOException {
...
 					1, new ExecutorThreadFactory("jdbc-upsert-output-format"));
 			this.scheduledFuture = this.scheduler.scheduleWithFixedDelay(() -> {
-				if (closed) {
-					return;
-				}
-				try {
-					flush();
-				} catch (Exception e) {
-					flushException = e;
+				synchronized (JDBCUpsertOutputFormat.this) {
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java]: Why should we add `synchronized`? `flush()` is already `synchronized`, right?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java [hunk 2]
```diff
@@ -184,15 +186,6 @@ public synchronized void close() throws IOException {
 		if (this.scheduledFuture != null) {
 			scheduledFuture.cancel(false);
 			this.scheduler.shutdown();
```
Comment (wuchong) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java]: Should we change this to `shutdownNow()`? I think the async flusher can quit now.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java (hunk 1 above)
Comment (JingsongLi) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java]: Make sure the flusher will graceful exit after close.

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java (hunk 2 above)
Comment (JingsongLi) [reply flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java]: In this case, I think we should not interrupt thread, we have good way to deal with it now, interrupt may make things complicate.

**Pull request merged**